includes/data-events/class-popups.php

```diff
@@ -180,18 +180,21 @@ public static function registration_submission_with_status( $email, $user_id, $p
 	 *
 	 * @param string         $email  Email address of the reader.
 	 * @param array|WP_Error $result Contact data if it was added, or error otherwise.
-	 * @param int|false      $popup_id The ID of the popup that triggered the registration, or false if not triggered by a popup.
+	 * @param array          $metadata Some metadata about the subscription. Always contains `current_page_url`, `newspack_popup_id` and `newsletters_subscription_method` keys.
 	 * @return ?array
 	 */
-	public static function newsletter_submission( $email, $result, $popup_id = false ) {
+	public static function newsletter_submission( $email, $result, $metadata = [] ) {
+		$popup_id = $metadata['newspack_popup_id'];
 		if ( ! $popup_id ) {
 			return;
 		}
 		$popup_data = self::get_popup_metadata( $popup_id );
 		return array_merge(
 			$popup_data,
 			[
-				'action' => self::FORM_SUBMISSION,
+				'action'                         => self::FORM_SUBMISSION,
+				'newsletter_subscription_method' => $metadata['newsletters_subscription_method'],
```

Comment:
For the sake of preserving the root object of `campaign_interaction` more agnostic and with permanent keys, I'd place this type of metadata nested in an isolated key since it's regarding a block inside the campaign. Something like `campaign_data` or `blocks_data`?

Something similar is done here, for donations:

https://github.com/Automattic/newspack-plugin/blob/aa5543a7c507b673650b0bb9ce9af0fd4a3965d3/includes/data-events/listeners.php#L219-L233

The platform may vary and have specific metadata associated with it. That shouldn't mess with the root object.

Reply:
Also, documenting in `README.md` also helps with perspective on how we're building these.

Reply:
> Something similar is done here, for donations:

IN donations all info about the donation is block specific as well. `amount`, `currency`, etc... It's the same case. If I change here I'd have to change there as well.

I get the point, can I just ask to do this refactor on another PR? because this will also touch the gha-4 connector and there's another PR open and they'll all conflict if I start this now.

> Also, documenting in README.md also helps with perspective on how we're building these.

I'm doing this! :) Since I'm touching this in several PRs and they are still changing a bit I'll do it all at once as one PR at the end.

GA4 connector will have its own README as well.

Reply:
> IN donations all info about the donation is block specific as well. amount, currency, etc... It's the same case. If I change here I'd have to change there as well.

I don't understand the relation. Every `donation_new` must have `amount`, `currency`, etc... It's a requirement of the action event regardless of the platform.

The `campaign_interaction` have some clear requirements, like `action` and `referer`. Other arguments are specific to the contents of the campaign, therefore not always there. If the event was called `campaign_interaction_donation` or `campaign_interaction_newsletter_subscription` it would be a different case.

We're not ensuring a good schema here.

Reply:
These listeners are not just to serve GA4, they are general purpose.

Reply:
ah, the other PR got merged. I can do this refactor now.

I can add `registration_data`, `donation_data` and `newsletter_subscription_data` keys in the root and place the values there. WDYT?

Reply:
`interaction_data` sounds like a good candidate as well for holding these

Reply:
why not an abstract one, so APIs don't have to struggle with it? if the action tells what it is, why would they key also have to?

Reply:
> interaction_data sounds like a good candidate as well for holding these

To ensure we always have the same schema. And to be consistent with `has_registration`, `has_newsletter` and `has_donation`.

I know it's unlikely, but in theory you can have more than one of these blocks inside a popup.

Also, we can know beforehand the schema inside each of these keys as well.

>  if the action tells what it is, why would the key also have to?

The action does not tell us that. At least not yet. that's being done by the `has_*` params